<a href="https://colab.research.google.com/github/Gen54cw/Analitica-de-Datos-Tintaya-Daniel/blob/main/laboratorios/LAB01_TintayaDaniel.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Laboratorio 1: Evaluación de factibilidad de un caso de uso analítico

Alumno: Tintaya Avila Daniel Eduardo

Codigo: 21200032

Fecha: 26/08/2026

# Paso 1. Configuración del entorno

In [22]:
import pandas as pd
import numpy as np
import unicodedata
import matplotlib.pyplot as plt
pd.set_option("display.max_columns", 50)
pd.set_option("display.width", 140)
print("pandas:", pd.__version__)
print("numpy :", np.__version__)


pandas: 2.2.3
numpy : 2.1.3


Las versiones utilizadas en el notebook corresponden a:

pandas: 2.2.3

numpy: 2.1.3

# Paso 2 · Carga e inventario inicial

In [23]:
#Para subir los csv
#from google.colab import files
#subidos = files.upload() # seleccione los dos archivos CSV
#print(list(subidos.keys()))

In [24]:
citas = pd.read_csv("citas_red_salud.csv", encoding="utf-8")
sedes = pd.read_csv("catalogo_sedes.csv", encoding="utf-8")
print("citas :", citas.shape)
print("sedes :", sedes.shape)
print()
citas.info()
display(citas.head(5))
display(sedes)

citas : (12200, 20)
sedes : (14, 6)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12200 entries, 0 to 12199
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id_cita                 12200 non-null  object 
 1   id_paciente             12200 non-null  object 
 2   sede                    12200 non-null  object 
 3   distrito                11467 non-null  object 
 4   especialidad            12055 non-null  object 
 5   fecha_programacion      12200 non-null  object 
 6   fecha_cita              12200 non-null  object 
 7   hora_cita               12200 non-null  object 
 8   canal_agendamiento      11652 non-null  object 
 9   dias_anticipacion       11933 non-null  float64
 10  edad_paciente           11854 non-null  float64
 11  citas_previas           12200 non-null  int64  
 12  inasistencias_previas   12200 non-null  int64  
 13  telefono_confirmado     4586 non-null   object 
 14  d

/usr/local/lib/python3.13/dist-packages/google/colab/_dataframe_summarizer.py:88: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  cast_date_col = pd.to_datetime(column, errors="coerce")


,id_cita,id_paciente,sede,distrito,especialidad,fecha_programacion,fecha_cita,hora_cita,canal_agendamiento,dias_anticipacion,edad_paciente,citas_previas,inasistencias_previas,telefono_confirmado,diagnostico_cie10,estado_cita,fecha_cancelacion,hora_registro_atencion,duracion_atencion_min,motivo_no_asistencia
0,C2407125,P146731,CAP San Martín de Porres,San Martín de Porres,Cardiología,2025-06-11,2025-07-15,09:15,App,34.0,24.0,7,2,NaN,L20.9,Cancelada,2025-06-26,NaN,NaN,NaN
1,C2400808,P122053,CAP Surquillo,Surquillo,Ginecología,2025-08-20,01/09/2025,08:00,Telefonico,12.0,66.0,2,2,Si,E78.5,Atendida,NaN,08:00,6.0,NaN
2,C2401689,P142968,CAP Independencia,Independencia,Dermatología,2026-01-07,2026-01-14,10:45,Presencial,7.0,33.0,4,1,No,G43.9,Cancelada,2026-01-08,NaN,NaN,NaN
3,C2410000,P101367,CAP Santa Anita,Santa Anita,Endocrinología,2025-10-30,2025-12-01,11:30,app movil,32.0,4.0,9,2,NaN,J45,Atendida,NaN,11:13,14.0,NaN
4,C2401784,P116873,CAP Puente Piedra,Puente Piedra,Odontología,2024-11-02,2024-11-14,16:15,Web,12.0,50.0,5,1,NaN,E11.9,Atendida,NaN,16:40,10.0,NaN


,sede,distrito,nivel_establecimiento,consultorios_ambulatorios,anio_apertura,horas_medico_semana
0,CAP San Juan de Lurigancho,San Juan de Lurigancho,I-4,22,2009,980
1,CAP Comas,Comas,I-4,18,2011,810
2,CAP Villa El Salvador,Villa El Salvador,I-3,14,2014,640
3,CAP Ate,Ate,I-4,19,2010,850
4,CAP San Martin de Porres,San Martín de Porres,I-3,15,2013,690
5,CAP Callao,Callao,II-1,26,2006,1180
6,CAP Los Olivos,Los Olivos,I-4,17,2012,760
7,CAP Chorrillos,Chorrillos,I-3,13,2015,590
8,CAP Santa Anita,Santa Anita,I-3,12,2016,545
9,CAP Puente Piedra,Puente Piedra,I-3,11,2017,500


**Preguntas**

**2.1 ¿Cuántas filas y cuántas columnas tiene cada archivo?**

citas:  12200 filas y 20 columnas

sedes: 14 filas y 6 columnas

2.2 ¿Qué columnas deberían ser numéricas o de fecha pero llegaron como texto? Enumérelas.


*   fecha_programacion
*   fecha_cita
*   hora_cita
*   fecha_cancelacion
*   hora_registro_atencion

Estas columnas no deben ser consideradas como texto


**2.3 A partir de los nombres de las columnas y de las primeras filas, ¿cuál cree que es la granularidad del archivo principal, es decir, qué representa una fila?**

La fila representa una cita programada de un paciente en una sede determinada, además de contener información relacionada con la programación de la cita, el paciente y el estado final de la cita.

# Paso 3 · Diccionario de datos

## Diccionario de datos

| Variable | Significado | Tipo esperado |
|---|---|---|
| `id_cita` | Identificador de la cita programada. Debería ser único. | Texto (clave) |
| `id_paciente` | Identificador del paciente. Se repite entre citas. | Texto (clave foránea) |
| `sede` | Establecimiento donde se programó la atención. | Categórica |
| `distrito` | Distrito del establecimiento. | Categórica |
| `especialidad` | Especialidad médica de la cita. | Categórica |
| `fecha_programacion` | Fecha en que se agendó la cita. | Fecha |
| `fecha_cita` | Fecha para la que se programó la atención. | Fecha |
| `hora_cita` | Hora programada de la atención. | Hora |
| `canal_agendamiento` | Medio por el que se agendó: presencial, teléfono, app o web. | Categórica |
| `dias_anticipacion` | Días entre el agendamiento y la cita, según el sistema. | Entero ≥ 0 |
| `edad_paciente` | Edad del paciente en años. | Entero, 0 a 110 |
| `citas_previas` | Número de citas históricas del paciente en la red. | Entero ≥ 0 |
| `inasistencias_previas` | Número de inasistencias históricas del paciente. | Entero ≥ 0 |
| `telefono_confirmado` | Si el teléfono de contacto fue verificado. | Categórica (Si/No) |
| `diagnostico_cie10` | Código de diagnóstico asociado a la atención. | Categórica |
| `estado_cita` | Estado final: Atendida, No atendida o Cancelada. | Categórica |
| `fecha_cancelacion` | Fecha en que el paciente canceló, si canceló. | Fecha |
| `hora_registro_atencion` | Hora en que se registró el ingreso del paciente al consultorio. | Hora |
| `duracion_atencion_min` | Duración de la consulta en minutos. | Entero |
| `motivo_no_asistencia` | Motivo registrado cuando el paciente no se presentó. | Categórica |

In [25]:
categoricas = ["sede", "distrito", "especialidad", "canal_agendamiento",
 "estado_cita", "telefono_confirmado"]
for col in categoricas:
 print("=" * 60)
 print(col, "->", citas[col].nunique(dropna=True), "valores distintos")
 print(citas[col].value_counts(dropna=False).head(12))

sede -> 30 valores distintos
sede
CAP Ate                       877
CAP Santa Anita               871
CAP San Martín de Porres      853
CAP Lince                     851
CAP El Agustino               848
CAP Villa El Salvador         838
CAP San Juan de Lurigancho    837
CAP Chorrillos                837
CAP Surquillo                 834
CAP Callao                    829
CAP Los Olivos                824
CAP Puente Piedra             819
Name: count, dtype: int64
distrito -> 14 valores distintos
distrito
Ate                       855
Lince                     839
Santa Anita               835
Villa El Salvador         832
San Martín de Porres      824
San Juan de Lurigancho    822
El Agustino               820
Chorrillos                815
Puente Piedra             813
Surquillo                 810
Callao                    806
Los Olivos                803
Name: count, dtype: int64
especialidad -> 30 valores distintos
especialidad
Odontología         1175
Oftalmología        1119
Neur

**RESPUESTA PASO 3**

**3.1 La red tiene catorce sedes y diez especialidades. ¿Cuántos valores distintos reporta el archivo para esas dos columnas? ¿Cómo se explica la diferencia?**

El archivo reporta 30 valores distintos para sede y 30 valores distintos para especialidad. La diferencia se explica en la forma de registrar las categorías. Por ejemplo, Dermatología y DERMATOLOGÍA aparecen como valores distintos. Por lo tanto, el número de valores distintos observado en el archivo no coincide

**3.2 Identifique al menos tres formas en que un mismo valor aparece escrito de manera distinta.**

Se observan, entre otras, las siguientes variantes:

*   Dermatología / DERMATOLOGÍA
*   Presencial / PRESENCIAL
*   App / app movil

**3.3 ¿Qué columna tiene una proporción de valores faltantes tan alta que probablemente no sea utilizable? Justifique.**

La columna de telefono_confirmado, porque de 12200 registros 7614 esta como NaN lo que indica que no contiene información

# Paso 4 · Normalización de categorías y fechas

In [26]:
def normaliza(texto):
    """Minúsculas, sin tildes, sin espacios sobrantes."""
    if pd.isna(texto):
        return np.nan

    t = str(texto).strip().lower()
    t = unicodedata.normalize("NFKD", t).encode("ascii", "ignore").decode("ascii")
    return " ".join(t.split())


for col in ["sede", "distrito", "especialidad", "canal_agendamiento", "estado_cita"]:
    citas[col + "_norm"] = citas[col].apply(normaliza)

print(citas["estado_cita_norm"].value_counts(dropna=False))
print()
print(citas["canal_agendamiento_norm"].value_counts(dropna=False))
print()
print("sedes distintas :", citas["sede_norm"].nunique())
print("especialidades distintas:", citas["especialidad_norm"].nunique())

estado_cita_norm
atendida       7252
no atendida    3652
cancelada      1296
Name: count, dtype: int64

canal_agendamiento_norm
presencial    4024
telefonico    2879
app           2445
web           2070
NaN            548
app movil      234
Name: count, dtype: int64

sedes distintas : 15
especialidades distintas: 10


In [27]:
def parse_fecha(serie):
    iso = pd.to_datetime(serie, format="%Y-%m-%d", errors="coerce")
    lat = pd.to_datetime(serie, format="%d/%m/%Y", errors="coerce")
    return iso.fillna(lat)


citas["fecha_cita_dt"] = parse_fecha(citas["fecha_cita"])
citas["fecha_prog_dt"] = parse_fecha(citas["fecha_programacion"])

print("fecha_cita no interpretadas :", citas["fecha_cita_dt"].isna().sum())
print("fecha_programacion no interpretadas :", citas["fecha_prog_dt"].isna().sum())

print(
    "rango de fechas de cita:",
    citas["fecha_cita_dt"].min().date(),
    "a",
    citas["fecha_cita_dt"].max().date()
)

fecha_cita no interpretadas : 0
fecha_programacion no interpretadas : 0
rango de fechas de cita: 2024-01-02 a 2026-06-29


**RESPUESTAS PASO 4**

**4.1 ¿Cuántos valores distintos quedaron en sede y especialidad después de normalizar? ¿Coincide ahora con lo que la red declara?**

Quedaron 15 sedes y 10 especialidades distintas. No, no coincide en las sedes, porque la red declara 14 sedes lo que indica que la normalización no logro capturar todo

**4.2 Después de normalizar, el canal de agendamiento todavía tiene una categoría de más. ¿Cuál es y qué haría usted con ella?**

La categoria que sobra es la de app movil que es la misma que app. Lo trataria como una variante de app, y las unificaria en una columna, incluyendo los registros faltantes

**4.3 ¿Por qué es una mala práctica sobrescribir la columna original en lugar de crear una columna nueva?**

Porque se pierde como fue registrado de forma original y no permitiria trazar y verificar la información.

# Paso 5 · Resolución de duplicados

In [28]:
# a) Duplicados exactos
dup_exactos = citas.duplicated().sum()
print("Filas idénticas en todas sus columnas:", dup_exactos)

# b) Identificadores repetidos con contenido distinto
citas_sin_exactos = citas.drop_duplicates()
mask = citas_sin_exactos["id_cita"].duplicated(keep=False)
print("Filas con id_cita repetido y contenido distinto:", mask.sum())

# Inspeccione un caso concreto antes de decidir
ejemplo = citas_sin_exactos.loc[mask, "id_cita"].iloc[0]
display(citas_sin_exactos[citas_sin_exactos["id_cita"] == ejemplo])


Filas idénticas en todas sus columnas: 127
Filas con id_cita repetido y contenido distinto: 146


,id_cita,id_paciente,sede,distrito,especialidad,fecha_programacion,fecha_cita,hora_cita,canal_agendamiento,dias_anticipacion,edad_paciente,citas_previas,inasistencias_previas,telefono_confirmado,diagnostico_cie10,estado_cita,fecha_cancelacion,hora_registro_atencion,duracion_atencion_min,motivo_no_asistencia,sede_norm,distrito_norm,especialidad_norm,canal_agendamiento_norm,estado_cita_norm,fecha_cita_dt,fecha_prog_dt
165,C2403465,P107703,CAP Ate,Ate,Medicina General,2026-02-04,2026-02-23,09:30,Web,19.0,1.0,5,2,Si,H52.1,No atendida,NaN,NaN,NaN,No registrado,cap ate,ate,medicina general,web,no atendida,2026-02-23,2026-02-04
4013,C2403465,P107703,CAP Ate,Ate,Medicina General,2026-02-04,2026-02-23,18:30,app movil,19.0,1.0,5,2,Si,H52.1,No atendida,NaN,NaN,NaN,No registrado,cap ate,ate,medicina general,app movil,no atendida,2026-02-23,2026-02-04


In [29]:
citas_lim = (citas
    .drop_duplicates() # elimina exactos
    .drop_duplicates(subset="id_cita", keep="first") # conserva primera versión
    .reset_index(drop=True))

print("Filas originales :", len(citas))
print("Filas resultantes:", len(citas_lim))
print("Filas descartadas:", len(citas) - len(citas_lim),
      f"({(len(citas)-len(citas_lim))/len(citas)*100:.2f} %)")


Filas originales : 12200
Filas resultantes: 12000
Filas descartadas: 200 (1.64 %)


**RESPUESTAS PASO 5**

5.1 ¿Cuántos duplicados exactos y cuántos identificadores repetidos encontró?

Hay 127 duplicados exactos y 146 filas con id_cita repetido y contenido distinto.

5.2 Observe el caso concreto que inspeccionó: ¿en qué columnas difieren las dos versiones de la misma cita?

Difieren en la hora_cita, canal_Agendamiento y canal_agendamiento_norm

5.3 Conservar la primera aparición es una decisión arbitraria. Proponga un criterio mejor para decidir cuál versión conservar y explique qué información adicional necesitaría para aplicarlo.

Se conservaria la versión más reciente, es decir la ultima en modificarse, para lo cual se necesitaria la fecha y hora de modificación del dataset

# Paso 6 · Auditoría de la variable objetivo

In [30]:
# Definición operativa de la variable objetivo
citas_lim["inasistencia"] = (citas_lim["estado_cita_norm"] == "noatendida").astype(int)
print(citas_lim["estado_cita_norm"].value_counts(dropna=False))
print()
print("Citas marcadas como inasistencia:", citas_lim["inasistencia"].sum())
# ¿Hay estados nulos o inesperados que no encajen en las tres categorías?
esperados = {"atendida", "no atendida", "cancelada"}
raros = set(citas_lim["estado_cita_norm"].dropna().unique()) - esperados
print("Estados inesperados:", raros if raros else "ninguno")
print("Estados nulos :", citas_lim["estado_cita_norm"].isna().sum())


estado_cita_norm
atendida       7128
no atendida    3592
cancelada      1280
Name: count, dtype: int64

Citas marcadas como inasistencia: 0
Estados inesperados: ninguno
Estados nulos : 0


In [31]:
coherencia = pd.crosstab(citas_lim["estado_cita_norm"],
                         citas_lim["fecha_cancelacion"].notna(),
                         rownames=["estado"], colnames=["tiene fecha_cancelacion"])
display(coherencia)

tiene fecha_cancelacion,False,True
estado,,
atendida,7098,30
cancelada,0,1280
no atendida,3567,25


**Preguntas a responder en el notebook (Paso 6)**

**6.1 ¿Qué proporción de las citas quedó clasificada en cada uno de los tres estados?**

Atendida: 7 128 (59,40 %)

No atendida: 3 592 (29,93 %)

Cancelada: 1 280 (10,67 %)


6.2 La tabla de coherencia muestra registros contradictorios. Descríbalos y cuantifíquelos.
**texto en negrita**
Se encontraron 55 registros contradictorios ya que 30 citas atendidas tienen fecha de cancelación y 25 citas no atendidas tienen fecha de cancelación.

**6.3 A la luz de lo anterior, ¿considera que la variable objetivo está registrada de forma suficientemente confiable como para entrenar un modelo? Responda sí o no y sustente en una o dos frases. Esta respuesta condiciona todo el resto del laboratorio.**

Si, porque a pesar de existir 55 registros contradictorios estos solo representan el 0.46% del total de 12000 registros. Demostrando que más del 99% es constistente y las contradicciones son detectables, lo que puede corregirse


# Paso 8 · Perfil de calidad de datos

In [32]:
#Completitud
perfil = pd.DataFrame({
    "nulos": citas_lim.isna().sum(),
    "nulos_%": (citas_lim.isna().mean() * 100).round(2),
    "unicos": citas_lim.nunique(dropna=True),
    "tipo": citas_lim.dtypes.astype(str)
}).sort_values("nulos_%", ascending=False)
display(perfil)

,nulos,nulos_%,unicos,tipo
fecha_cancelacion,10665,88.88,721,object
motivo_no_asistencia,8441,70.34,5,object
telefono_confirmado,7504,62.53,2,object
hora_registro_atencion,4894,40.78,648,object
duracion_atencion_min,4894,40.78,36,float64
distrito,720,6.00,14,object
distrito_norm,720,6.00,14,object
canal_agendamiento_norm,538,4.48,5,object
canal_agendamiento,538,4.48,6,object
edad_paciente,335,2.79,102,float64


In [35]:
#8.2 Validez
edad_mala = ((citas_lim["edad_paciente"] < 0) |
 (citas_lim["edad_paciente"] > 110)).sum()
antic_mala = ((citas_lim["dias_anticipacion"] < 0) |
 (citas_lim["dias_anticipacion"] > 365)).sum()
print("Edades fuera de dominio :", edad_mala)
print("Anticipaciones fuera de dominio :", antic_mala)
print()
display(citas_lim.loc[(citas_lim["edad_paciente"] < 0) |
 (citas_lim["edad_paciente"] > 110),["id_cita", "edad_paciente", "dias_anticipacion"]].head(10))

Edades fuera de dominio : 44
Anticipaciones fuera de dominio : 35



,id_cita,edad_paciente,dias_anticipacion
93,C2406150,201.0,17.0
711,C2406489,155.0,5.0
1006,C2403059,-1.0,14.0
1180,C2400968,155.0,17.0
1232,C2407695,132.0,19.0
1511,C2410150,-3.0,9.0
1628,C2401562,155.0,2.0
2123,C2402036,201.0,10.0
2299,C2400373,132.0,NaN
2604,C2401370,-3.0,10.0


In [34]:
#8.3 Consistencia interna
citas_lim["antic_calculada"] = (citas_lim["fecha_cita_dt"] - citas_lim["fecha_prog_dt"]).dt.days
dif = (citas_lim["antic_calculada"] - citas_lim["dias_anticipacion"]).abs()
inconsistentes = (dif > 1).sum()
print("Registros donde dias_anticipacion no coincide con las fechas:", inconsistentes)
print(f"Equivale al {inconsistentes/len(citas_lim)*100:.2f} % del total")
print()
display(citas_lim.loc[dif > 1, ["id_cita", "fecha_prog_dt", "fecha_cita_dt", "dias_anticipacion", "antic_calculada"]].head(8))


Registros donde dias_anticipacion no coincide con las fechas: 241
Equivale al 2.01 % del total



,id_cita,fecha_prog_dt,fecha_cita_dt,dias_anticipacion,antic_calculada
111,C2406416,2025-10-16,2025-11-01,23.0,16
130,C2408203,2024-10-13,2024-10-22,33.0,9
137,C2408078,2024-05-31,2024-06-17,999.0,17
232,C2402378,2025-09-02,2025-09-22,28.0,20
291,C2401823,2024-09-14,2024-09-16,5.0,2
293,C2402617,2024-10-05,2024-10-14,39.0,9
422,C2407976,2025-09-14,2025-09-22,46.0,8
449,C2409202,2025-02-25,2025-03-05,42.0,8
